crates/environments/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+    envs: &[TargetEnvironment],
+    trigger_type: &TriggerType,
+    component: &ComponentToValidate<'_>,
+) -> anyhow::Result<Vec<anyhow::Error>> {
+    let mut errs = vec![];
+
+    for env in envs {
+        let worlds = env.worlds(trigger_type);
+        if let Some(e) = validate_wasm_against_any_world(env, &worlds, component)
+            .await
+            .err()
+        {
+            errs.push(e);
+        }
+    }
+
+    if errs.is_empty() {
+        tracing::info!(
+            "Validated component {} {} against all target worlds",
+            component.id(),
+            component.source_description()
+        );
+    }
+
+    Ok(errs)
+}
+
+async fn validate_wasm_against_any_world(
+    env: &TargetEnvironment,
+    world_names: &[String],
+    component: &ComponentToValidate<'_>,
+) -> anyhow::Result<()> {
+    let mut result = Ok(());
+    for target_world in world_names {
+        tracing::debug!(
+            "Trying component {} {} against target world {target_world}",
+            component.id(),
+            component.source_description(),
+        );
+        match validate_wasm_against_world(env, target_world, component).await {
+            Ok(()) => {
+                tracing::info!(
+                    "Validated component {} {} against target world {target_world}",
+                    component.id(),
+                    component.source_description(),
+                );
+                return Ok(());
+            }
+            Err(e) => {
+                // Record the error, but continue in case a different world succeeds
+                tracing::info!(
+                    "Rejecting component {} {} for target world {target_world} because {e:?}",
+                    component.id(),
+                    component.source_description(),
+                );
+                result = Err(e);
+            }
+        }
+    }
+    result
+}
+
+async fn validate_wasm_against_world(
+    env: &TargetEnvironment,
+    target_world: &str,
+    component: &ComponentToValidate<'_>,
+) -> anyhow::Result<()> {
+    // Because we are abusing a composition tool to do validation, we have to
+    // provide a name by which to refer to the component in the dummy composition.
+    let component_name = "root:component";
+    let component_key = wac_types::BorrowedPackageKey::from_name_and_version(component_name, None);
+
+    // wac is going to get the world from the environment package bytes.
+    // This constructs a key for that mapping.
+    let env_pkg_name = env.package_namespaced_name();
+    let env_pkg_key =
+        wac_types::BorrowedPackageKey::from_name_and_version(&env_pkg_name, env.package_version());
+
+    let env_name = env.name();
+
+    let wac_text = format!(
+        r#"
+    package validate:component@1.0.0 targets {target_world};
+    let c = new {component_name} {{ ... }};
+    export c...;
+    "#
+    );
+
+    let doc = wac_parser::Document::parse(&wac_text)
```

Comment:
We really should implement a programatic API in `wac` for this so that we don't need to manipulate a wac script.

Reply:
Yeah!  I tried constructing a wac document AST, which seemed cleaner than parsing a document, but I foundered on - if memory serves - something as trivial as the `miette` spans.  I tried it with dummy spans but something internal seems to have been trying to use the length of the name in error or something and anyway there was anguish.  Using the underlying model would have been awesome, but was (for me) hard to understand and to line up inputs for.